shotover-proxy/src/transforms/throttling.rs

```diff
@@ -0,0 +1,73 @@
+use crate::{
+    error::ChainResponse,
+    transforms::{Transform, Transforms, Wrapper},
+};
+use anyhow::Result;
+use async_trait::async_trait;
+use governor::{
+    clock::DefaultClock,
+    middleware::NoOpMiddleware,
+    state::{InMemoryState, NotKeyed},
+    Quota, RateLimiter,
+};
+use serde::Deserialize;
+use std::num::NonZeroU32;
+use std::sync::Arc;
+
+#[derive(Deserialize, Debug, Clone)]
+pub struct RequestThrottlingConfig {
+    pub max_requests_per_second: NonZeroU32,
+}
+
+impl RequestThrottlingConfig {
+    pub async fn get_source(&self) -> Result<Transforms> {
+        tracing::info!("here");
+        Ok(Transforms::RequestThrottling(RequestThrottling {
+            limiter: Arc::new(RateLimiter::direct(Quota::per_second(
+                self.max_requests_per_second,
+            ))),
+        }))
+    }
+}
+
+#[derive(Clone)]
+pub struct RequestThrottling {
+    limiter: Arc<RateLimiter<NotKeyed, InMemoryState, DefaultClock, NoOpMiddleware>>,
+}
+
+#[async_trait]
+impl Transform for RequestThrottling {
```

Comment:
I came up with a way of writing this method that should help reduce allocations in 2 ways:
1. We only allocate one vec per call now.
2. The iterator logic should preallocate the vec at the correct length without resizes. (The itererator chaining is a little complicated so its possible rust might not be able to perform this optimization here, would have to check the assembly if we needed to know for sure)


```rust
#[async_trait]
impl Transform for RequestThrottling {
    async fn transform<'a>(&'a mut self, mut message_wrapper: Wrapper<'a>) -> ChainResponse {
        // extract throttled message from the message_wrapper
        let throttled_messages: Vec<(Message, usize)> = (0..message_wrapper.messages.len())
            .into_iter()
            .filter(|_| self.limiter.check().is_err())
            .map(|i| {
                let message = message_wrapper.messages.remove(i);
                (message, i)
            })
            .collect();

        let mut responses = message_wrapper.call_next_transform().await?;

        // reinsert into responses
        for (mut message, i) in throttled_messages {
            message.set_backpressure()?;
            responses.insert(i, message);
        }

        Ok(responses)
    }
}
```

Reply:
Thanks!